libraries/lib-utility/MessageBuffer.h

```diff
@@ -0,0 +1,112 @@
+/**********************************************************************
+ 
+ Audacity: A Digital Audio Editor
+ 
+ @file MessageBuffer.h
+ 
+ Paul Licameli split from PlaybackSchedule.h
+ 
+ **********************************************************************/
+
+#ifndef __AUDACITY_MESSAGE_BUFFER__
+#define __AUDACITY_MESSAGE_BUFFER__
+
+//! Communicate data atomically from one writer thread to one reader.
+/*!
+ This is not a queue: it is not necessary for each write to be read.
+ Rather loss of a message is allowed:  writer may overwrite.
+ Data type must be default-constructible and reassignable.
+
+ Communication is lock-free for both threads, and also wait-free for the reader.
+ */
+template<typename Data>
+class MessageBuffer {
+   struct UpdateSlot {
+      Data mData;
+   };
+   NonInterfering<UpdateSlot> mSlots[2];
+
+   using Control = unsigned char;
+   //! Bit fields
+   enum : Control {
+      SlotToRead = 0x1, //!< Next slot to read, unless switched
+      SwitchSlot = 0x2, //!< Set only by writer; true only while not writing
+      ReadInProgress = 0x4, //!< Changed only by reader thread
+   };
+   static size_t ChooseReaderSlot(Control control)
```

Comment:
A very complex function to understand. I think it is worth commenting on its meaning or rewriting it without bit magic.

Also, the choice of types is confusing as well.